File: npm-packages/meteor-capacitor/ios/Sources/MeteorjsCapacitor/WebAppConfiguration.swift
// Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
// Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.

//
// WebAppConfiguration.swift
//
// Manages configuration and settings for the Meteor webapp, including
// persistence of app state and runtime configuration values.
//
// This file is copied directly from cordova-plugin-meteor-webapp with
// minimal changes to adapt for Capacitor.
//

import Foundation

final class WebAppConfiguration {
    let userDefaults: UserDefaults

    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
    }

    /// The appId as defined in the runtime config
    var appId: String? {
        get {
            return userDefaults.string(forKey: "MeteorWebAppId")
        }
        set {
            let oldValue = appId
            if newValue != oldValue {
                if oldValue != nil {
                    NSLog("appId seems to have changed, new: \(String(describing: newValue)), old: \(oldValue!)")
                }
                if let newValue {
                    userDefaults.set(newValue, forKey: "MeteorWebAppId")
                } else {
                    userDefaults.removeObject(forKey: "MeteorWebAppId")
                }
            }
        }
    }

    /// The rootURL as defined in the runtime config
    var rootURL: URL? {
        get {
            return userDefaults.url(forKey: "MeteorWebAppRootURL")
        }
        set {
            let oldValue = rootURL
            if newValue != oldValue {
                if oldValue != nil {
                    NSLog("ROOT_URL seems to have changed, new: \(String(describing: newValue)), old: \(oldValue!)")
                }
                if let newValue {
                    userDefaults.set(newValue, forKey: "MeteorWebAppRootURL")
                } else {
                    userDefaults.removeObject(forKey: "MeteorWebAppRootURL")
                }
            }
        }
    }

    /// The Cordova compatibility version as specified in the asset manifest
    var cordovaCompatibilityVersion: String? {
        get {
            return userDefaults.string(forKey: "MeteorWebAppCordovaCompatibilityVersion")
        }

        set {
            if newValue != cordovaCompatibilityVersion {
                if newValue == nil {
                    userDefaults.removeObject(forKey: "MeteorWebAppCordovaCompatibilityVersion")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppCordovaCompatibilityVersion")
                }
            }
        }
    }

    /// The last seen initial version of the asset bundle
    var lastSeenInitialVersion: String? {
        get {
            return userDefaults.string(forKey: "MeteorWebAppLastSeenInitialVersion")
        }

        set {
            if newValue != lastSeenInitialVersion {
                if newValue == nil {
                    userDefaults.removeObject(forKey: "MeteorWebAppLastSeenInitialVersion")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppLastSeenInitialVersion")
                }
            }
        }
    }

    /// The last downloaded version of the asset bundle
    var lastDownloadedVersion: String? {
        get {
            return userDefaults.string(forKey: "MeteorWebAppLastDownloadedVersion")
        }

        set {
            if newValue != lastDownloadedVersion {
                if newValue == nil {
                    userDefaults.removeObject(forKey: "MeteorWebAppLastDownloadedVersion")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppLastDownloadedVersion")
                }
            }
        }
    }

    /// The last kwown good version of the asset bundle
    var lastKnownGoodVersion: String? {
        get {
            return userDefaults.string(forKey: "MeteorWebAppLastKnownGoodVersion")
        }

        set {
            if newValue != lastKnownGoodVersion {
                if newValue == nil {
                    userDefaults.removeObject(forKey: "MeteorWebAppLastKnownGoodVersion")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppLastKnownGoodVersion")
                }
            }
        }
    }

    /// Blacklisted asset bundle versions
    var blacklistedVersions: [String] {
        get {
            let versions =
                userDefaults.array(forKey: "MeteorWebAppBlacklistedVersions") as? [String] ?? []
            if !versions.isEmpty {
                NSLog("BLACKLIST - blacklistedVersions: \(versions)")
            }
            return versions
        }

        set {
            if newValue != blacklistedVersions {
                if newValue.isEmpty {
                    NSLog("BLACKLIST - removing blacklisted versions")
                    userDefaults.removeObject(forKey: "MeteorWebAppBlacklistedVersions")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppBlacklistedVersions")
                }
            }
        }
    }

    var versionsToRetry: [String] {
        get {
            let versions = userDefaults.array(forKey: "MeteorWebAppVersionsToRetry") as? [String] ?? []
            if !versions.isEmpty {
                NSLog("BLACKLIST - versionsToRetry: \(versions)")
            }
            return versions
        }
        set {
            if newValue != versionsToRetry {
                if newValue.isEmpty {
                    NSLog("BLACKLIST - removing versions to retry")
                    userDefaults.removeObject(forKey: "MeteorWebAppVersionsToRetry")
                } else {
                    userDefaults.set(newValue, forKey: "MeteorWebAppVersionsToRetry")
                }
            }
        }
    }

    func addBlacklistedVersion(_ version: String) {
        var blacklistedVersions = self.blacklistedVersions
        var versionsToRetry = self.versionsToRetry

        if !versionsToRetry.contains(version) && !blacklistedVersions.contains(version) {
            NSLog("BLACKLIST - adding faulty version to retry: \(version)")
            versionsToRetry.append(version)
            self.versionsToRetry = versionsToRetry
        } else {
            if let index = versionsToRetry.firstIndex(of: version) {
                versionsToRetry.remove(at: index)
                self.versionsToRetry = versionsToRetry
            }
            if !blacklistedVersions.contains(version) {
                blacklistedVersions.append(version)
                NSLog("BLACKLIST - blacklisting version: \(version)")
                self.blacklistedVersions = blacklistedVersions
            }
        }
    }

    func reset() {
        appId = nil
        rootURL = nil
        cordovaCompatibilityVersion = nil
        lastSeenInitialVersion = nil
        lastDownloadedVersion = nil
        lastKnownGoodVersion = nil
        blacklistedVersions = []
        versionsToRetry = []
    }
}

File: npm-packages/meteor-capacitor/android/src/main/java/com/meteorjs/capacitor/WebAppConfiguration.java
/*
 * Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
 * Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.
 */

package com.meteorjs.capacitor;

import android.content.SharedPreferences;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

final class WebAppConfiguration {

    private static final String KEY_APP_ID = "MeteorWebAppId";
    private static final String KEY_ROOT_URL = "MeteorWebAppRootURL";
    private static final String KEY_CORDOVA_COMPATIBILITY_VERSION = "MeteorWebAppCordovaCompatibilityVersion";
    private static final String KEY_LAST_DOWNLOADED_VERSION = "MeteorWebAppLastDownloadedVersion";
    private static final String KEY_LAST_SEEN_INITIAL_VERSION = "MeteorWebAppLastSeenInitialVersion";
    private static final String KEY_LAST_KNOWN_GOOD_VERSION = "MeteorWebAppLastKnownGoodVersion";
    private static final String KEY_BLACKLISTED_VERSIONS = "MeteorWebAppBlacklistedVersions";
    private static final String KEY_VERSIONS_TO_RETRY = "MeteorWebAppVersionsToRetry";

    private final SharedPreferences preferences;

    WebAppConfiguration(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    String getAppId() {
        return preferences.getString(KEY_APP_ID, null);
    }

    void setAppId(String appId) {
        setString(KEY_APP_ID, appId);
    }

    String getRootUrlString() {
        return preferences.getString(KEY_ROOT_URL, null);
    }

    void setRootUrlString(String rootUrlString) {
        setString(KEY_ROOT_URL, rootUrlString);
    }

    String getCordovaCompatibilityVersion() {
        return preferences.getString(KEY_CORDOVA_COMPATIBILITY_VERSION, null);
    }

    void setCordovaCompatibilityVersion(String version) {
        setString(KEY_CORDOVA_COMPATIBILITY_VERSION, version);
    }

    String getLastDownloadedVersion() {
        return preferences.getString(KEY_LAST_DOWNLOADED_VERSION, null);
    }

    void setLastDownloadedVersion(String version) {
        setString(KEY_LAST_DOWNLOADED_VERSION, version);
    }

    String getLastSeenInitialVersion() {
        return preferences.getString(KEY_LAST_SEEN_INITIAL_VERSION, null);
    }

    void setLastSeenInitialVersion(String version) {
        setString(KEY_LAST_SEEN_INITIAL_VERSION, version);
    }

    String getLastKnownGoodVersion() {
        return preferences.getString(KEY_LAST_KNOWN_GOOD_VERSION, null);
    }

    void setLastKnownGoodVersion(String version) {
        setString(KEY_LAST_KNOWN_GOOD_VERSION, version);
    }

    Set<String> getBlacklistedVersions() {
        Set<String> versions = preferences.getStringSet(KEY_BLACKLISTED_VERSIONS, Collections.emptySet());
        return new HashSet<>(versions);
    }

    Set<String> getVersionsToRetry() {
        Set<String> versions = preferences.getStringSet(KEY_VERSIONS_TO_RETRY, Collections.emptySet());
        return new HashSet<>(versions);
    }

    void addBlacklistedVersion(String version) {
        Set<String> blacklistedVersions = getBlacklistedVersions();
        Set<String> versionsToRetry = getVersionsToRetry();

        SharedPreferences.Editor editor = preferences.edit();
        if (!versionsToRetry.contains(version) && !blacklistedVersions.contains(version)) {
            versionsToRetry.add(version);
            editor.putStringSet(KEY_VERSIONS_TO_RETRY, versionsToRetry);
            editor.apply();
            return;
        }

        versionsToRetry.remove(version);
        blacklistedVersions.add(version);

        if (versionsToRetry.isEmpty()) {
            editor.remove(KEY_VERSIONS_TO_RETRY);
        } else {
            editor.putStringSet(KEY_VERSIONS_TO_RETRY, versionsToRetry);
        }

        editor.putStringSet(KEY_BLACKLISTED_VERSIONS, blacklistedVersions);
        editor.apply();
    }

    void reset() {
        preferences.edit().clear().apply();
    }

    private void setString(String key, String value) {
        SharedPreferences.Editor editor = preferences.edit();
        if (value == null) {
            editor.remove(key);
        } else {
            editor.putString(key, value);
        }
        editor.apply();
    }
}
